Looking at the test failures, I now see three kinds

```
11: ======================================================================
11: FAIL: testCreationEncodeAsNull (test_message.CodecTest)
11: ----------------------------------------------------------------------
11: Traceback (most recent call last):
11:   File "/home/runner/work/qpid-proton/qpid-proton/python/tests/proton_tests/test_message.py", line 267, in testCreationEncodeAsNull
11:     assert dproperties.descriptor == 0x73, (dproperties.descriptor)
11: AssertionError: 112
```

```
11: ======================================================================
91
11: FAIL: testUserId (test_message.AccessorsTest)
92
11: ----------------------------------------------------------------------
93
11: Traceback (most recent call last):
94
11:   File "/home/runner/work/qpid-proton/qpid-proton/python/tests/proton_tests/test_message.py", line 82, in testUserId
95
11:     self._test("user_id", b"", (b"asdf", b"fdsa", b"asd\x00fdsa", b""))
96
11:   File "/home/runner/work/qpid-proton/qpid-proton/python/tests/proton_tests/test_message.py", line 44, in _test
97
11:     assert d == default, (d, default)
98
11: AssertionError: (<cdata 'struct pn_bytes_t' owning 16 bytes>, b'')
```

```
11: ======================================================================
67
11: ERROR: testDefaultCreationExpiryDecode (test_message.CodecTest)
68
11: ----------------------------------------------------------------------
69
11: Traceback (most recent call last):
70
11:   File "/home/runner/work/qpid-proton/qpid-proton/python/tests/proton_tests/test_message.py", line 319, in testDefaultCreationExpiryDecode
71
11:     msg2.decode(bytes, size)
72
11:   File "/home/runner/work/qpid-proton/qpid-proton/python/proton/_message.py", line 531, in decode
73
11:     self._check(pn_message_decode(self._msg, bytes, size[0]))
74
11:   File "/home/runner/work/qpid-proton/qpid-proton/python/proton/_message.py", line 103, in _check
75
11:     raise exc("[%s]: %s" % (err, pn_error_text(pn_message_error(self._msg))))
76
11: proton._exceptions.MessageException: [-4]: <cdata 'char *' 0x55e0cda304c0>
```

--- a/python/proton/_data.py
+++ b/python/proton/_data.py
@@ -684,7 +684,11 @@ class Data:
     def _check(self, err):
         if err < 0:
             exc = EXCEPTIONS.get(err, DataException)
-            raise exc("[%s]: %s" % (err, pn_error_text(pn_data_error(self._data))))
+            raise exc("[%s]: %s" % (
+                err,
+                ffi.string(pn_error_text(pn_data_error(self._data))).decode('utf8')
+            )
+        )
         else:
             return err
 
@@ -816,7 +820,7 @@ class Data:
             else:
                 self._check(encoded_size)
 
-    def decode(self, bytes, size):
+    def decode(self, data):
         """
         Decodes the first value from supplied AMQP data and returns the
         number of bytes consumed.
@@ -825,6 +829,8 @@ class Data:
         :param encoded: AMQP encoded binary data
         :raise: :exc:`DataException` if there is a Proton error.
         """
+        bytes = ffi.new('char []', data)
+        size = ffi.new('size_t *', len(data))
         return self._check(pn_data_decode(self._data, bytes, size[0]))
 
     def put_list(self):

--- a/python/proton/_message.py
+++ b/python/proton/_message.py
@@ -44,7 +44,7 @@ from _proton_core.lib import PN_DEFAULT_PRIORITY, PN_OVERFLOW, pn_error_text, pn
     pn_message_set_content_type, pn_message_set_creation_time, pn_message_set_delivery_count, pn_message_set_durable, \
     pn_message_set_expiry_time, pn_message_set_first_acquirer, pn_message_set_group_id, pn_message_set_group_sequence, \
     pn_message_set_inferred, pn_message_set_priority, pn_message_set_reply_to, pn_message_set_reply_to_group_id, \
-    pn_message_set_subject, pn_message_set_ttl, pn_message_set_user_id
+    pn_message_set_subject, pn_message_set_ttl, pn_message_set_user_id, pn_bytes
 
 
 
@@ -100,7 +100,10 @@ class Message(object):
     def _check(self, err):
         if err < 0:
             exc = EXCEPTIONS.get(err, MessageException)
-            raise exc("[%s]: %s" % (err, pn_error_text(pn_message_error(self._msg))))
+            raise exc("[%s]: %s" % (
+                err,
+                ffi.string(pn_error_text(pn_message_error(self._msg))).decode('utf8')
+        ))
         else:
             return err
 
@@ -309,10 +312,14 @@ class Message(object):
         """)
 
     def _get_user_id(self):
-        return pn_message_get_user_id(self._msg)
+        r = pn_message_get_user_id(self._msg)
+        if r.start == ffi.NULL:
+            return b""
+        return ffi.buffer(r.start, r.size)[:]
 
     def _set_user_id(self, value):
-        self._check(pn_message_set_user_id(self._msg, value))
+        pnbytes = pn_bytes(len(value), value)
+        self._check(pn_message_set_user_id(self._msg, pnbytes))
 
     user_id = property(_get_user_id, _set_user_id, doc="""
         The user id of the message creator.
@@ -525,9 +532,11 @@ class Message(object):
                 continue
             else:
                 self._check(status_code)
-                return (bytes, size)
+                return ffi.buffer(bytes, size[0])[:]
 
-    def decode(self, bytes, size):
+    def decode(self, data):
+        bytes = ffi.new('char []', data)
+        size = ffi.new("size_t *", len(data))
         self._check(pn_message_decode(self._msg, bytes, size[0]))
         self._post_decode()
 

--- a/python/tests/proton_tests/test_message.py
+++ b/python/tests/proton_tests/test_message.py
@@ -46,7 +46,7 @@ class AccessorsTest(Test):
             setattr(self.msg, name, v)
             gotten = getattr(self.msg, name)
 
-            assert gotten == v, gotten
+            assert gotten == v, (gotten, v)
 
     def _test_symbol(self, name):
         self._test(name, symbol(None), (symbol(u"abc.123.#$%"), symbol(u"hello.world")))
@@ -127,10 +127,10 @@ class CodecTest(Test):
     def testProperties(self):
         self.msg.properties = {}
         self.msg.properties['key'] = 'value'
-        bytes, size = self.msg.encode()
+        data = self.msg.encode()
 
         msg2 = Message()
-        msg2.decode(bytes, size)
+        msg2.decode(data)
 
         assert msg2.properties['key'] == 'value', msg2.properties['key']
 
@@ -140,10 +140,10 @@ class CodecTest(Test):
 
     def testStringSubclassPropertyKey(self):
         self.msg.properties = {'abc': 123, CodecTest.MyStringSubclass('def'): 456}
-        bytes, size = self.msg.encode()
+        data = self.msg.encode()
 
         msg2 = Message()
-        msg2.decode(bytes, size)
+        msg2.decode(data)
 
         assert msg2.properties == self.msg.properties
         for k in msg2.properties:
@@ -203,10 +203,9 @@ class CodecTest(Test):
 
     def testAnnotationsSymbolicAndUlongKey(self, a={symbol('one'): 1, 'two': 2, ulong(3): 'three'}):
         self.msg.annotations = a
-        bytes, size = self.msg.encode()
-
+        data = self.msg.encode()
         msg2 = Message()
-        msg2.decode(bytes, size)
+        msg2.decode(data)
         # both keys must be symbols
         assert msg2.annotations == a
 
@@ -219,10 +218,10 @@ class CodecTest(Test):
         self.msg.subject = "subject"
         self.msg.body = 'Hello World!'
 
-        bytes, size = self.msg.encode()
+        data = self.msg.encode()
 
         msg2 = Message()
-        msg2.decode(bytes, size)
+        msg2.decode(data)
 
         assert self.msg.id == msg2.id, (self.msg.id, msg2.id)
         assert self.msg.correlation_id == msg2.correlation_id, (self.msg.correlation_id, msg2.correlation_id)
@@ -234,16 +233,16 @@ class CodecTest(Test):
 
     def testExpiryEncodeAsNull(self):
         self.msg.group_id = "A"  # Force creation and expiry fields to be present
-        bytes, size = self.msg.encode()
+        data = self.msg.encode()
 
         decoder = Data()
 
         # Skip past the headers
-        consumed = decoder.decode(bytes, size)
+        consumed = decoder.decode(data)
         decoder.clear()
-        # data = data[consumed:]
+        data = data[consumed:]
 
-        decoder.decode(bytes, size)
+        decoder.decode(data)
         dproperties = decoder.get_py_described()
         # Check we've got the correct described list
         assert dproperties.descriptor == 0x73, (dproperties.descriptor)
@@ -253,15 +252,13 @@ class CodecTest(Test):
 
     def testCreationEncodeAsNull(self):
         self.msg.group_id = "A"  # Force creation and expiry fields to be present
-        bytes, size = self.msg.encode()
+        data = self.msg.encode()
         decoder = Data()
-
         # Skip past the headers
-        consumed = decoder.decode(bytes, size)
+        consumed = decoder.decode(data)
         decoder.clear()
-        # data = data[consumed:]
-
-        decoder.decode(bytes, size)
+        data = data[consumed:]
+        decoder.decode(data)
         dproperties = decoder.get_py_described()
         # Check we've got the correct described list
         assert dproperties.descriptor == 0x73, (dproperties.descriptor)
@@ -271,16 +268,16 @@ class CodecTest(Test):
 
     def testGroupSequenceEncodeAsNull(self):
         self.msg.reply_to_group_id = "R"  # Force group_id and group_sequence fields to be present
-        bytes, size = self.msg.encode()
+        data = self.msg.encode()
 
         decoder = Data()
 
         # Skip past the headers
-        consumed = decoder.decode(bytes, size)
+        consumed = decoder.decode(data)
         decoder.clear()
-        # data = data[consumed:]
+        data = data[consumed:]
 
-        decoder.decode(bytes, size)
+        decoder.decode(data)
         dproperties = decoder.get_py_described()
         # Check we've got the correct described list
         assert dproperties.descriptor == 0x73, (dproperties.descriptor)
@@ -292,16 +289,16 @@ class CodecTest(Test):
     def testGroupSequenceEncodeAsNonNull(self):
         self.msg.group_id = "G"
         self.msg.reply_to_group_id = "R"  # Force group_id and group_sequence fields to be present
-        bytes, size = self.msg.encode()
+        data = self.msg.encode()
 
         decoder = Data()
 
         # Skip past the headers
-        consumed = decoder.decode(bytes, size)
+        consumed = decoder.decode(data)
         decoder.clear()
-        # data = data[consumed:]
+        data = data[consumed:]
 
-        decoder.decode(bytes, size)
+        decoder.decode(data)
         dproperties = decoder.get_py_described()
         # Check we've got the correct described list
         assert dproperties.descriptor == 0x73, (dproperties.descriptor)
@@ -313,38 +310,33 @@ class CodecTest(Test):
     def testDefaultCreationExpiryDecode(self):
         # This is a message with everything filled explicitly as null or zero in LIST32 HEADER and PROPERTIES lists
         data = b'\x00\x53\x70\xd0\x00\x00\x00\x0a\x00\x00\x00\x05\x42\x40\x40\x42\x52\x00\x00\x53\x73\xd0\x00\x00\x00\x12\x00\x00\x00\x0d\x40\x40\x40\x40\x40\x40\x40\x40\x40\x40\x40\x52\x00\x40'
-        bytes = ffi.new('char []', data)
-        size = ffi.new('size_t *', ffi.sizeof(bytes))
         msg2 = Message()
-        msg2.decode(bytes, size)
+        msg2.decode(data)
         assert msg2.expiry_time == 0, (msg2.expiry_time)
         assert msg2.creation_time == 0, (msg2.creation_time)
 
         # The same message with LIST8s instead
         data = b'\x00\x53\x70\xc0\x07\x05\x42\x40\x40\x42\x52\x00\x00\x53\x73\xc0\x0f\x0d\x40\x40\x40\x40\x40\x40\x40\x40\x40\x40\x40\x52\x00\x40'
-        bytes = ffi.new('char []', data)
-        size = ffi.new('size_t *', ffi.sizeof(bytes))
+
         msg3 = Message()
-        msg3.decode(bytes, size)
+        msg3.decode(data)
         assert msg2.expiry_time == 0, (msg2.expiry_time)
         assert msg2.creation_time == 0, (msg2.creation_time)
 
         # Minified message with zero length HEADER and PROPERTIES lists
         data = b'\x00\x53\x70\x45' b'\x00\x53\x73\x45'
-        bytes = ffi.new('char []', data)
-        size = ffi.new('size_t *', ffi.sizeof(bytes))
         msg4 = Message()
-        msg4.decode(bytes, size)
+        msg4.decode(data)
         assert msg2.expiry_time == 0, (msg2.expiry_time)
         assert msg2.creation_time == 0, (msg2.creation_time)
 
     def testDefaultPriorityEncode(self):
         assert self.msg.priority == 4, (self.msg.priority)
         self.msg.ttl = 0.003  # field after priority, so forces priority to be present
-        bytes, size = self.msg.encode()
+        data = self.msg.encode()
 
         decoder = Data()
-        decoder.decode(bytes, size)
+        decoder.decode(data)
 
         dheaders = decoder.get_py_described()
         # Check we've got the correct described list
@@ -357,24 +349,19 @@ class CodecTest(Test):
     def testDefaultPriorityDecode(self):
         # This is a message with everything filled explicitly as null or zero in LIST32 HEADER and PROPERTIES lists
         data = b'\x00\x53\x70\xd0\x00\x00\x00\x0a\x00\x00\x00\x05\x42\x40\x40\x42\x52\x00\x00\x53\x73\xd0\x00\x00\x00\x22\x00\x00\x00\x0d\x40\x40\x40\x40\x40\x40\x40\x40\x83\x00\x00\x00\x00\x00\x00\x00\x00\x83\x00\x00\x00\x00\x00\x00\x00\x00\x40\x52\x00\x40'
-        bytes = ffi.new('char []', data)
-        size = ffi.new('size_t *', ffi.sizeof(bytes))
         msg2 = Message()
-        msg2.decode(bytes, size)
+        msg2.decode(data)
         assert msg2.priority == 4, (msg2.priority)
 
         # The same message with LIST8s instead
         data = b'\x00\x53\x70\xc0\x07\x05\x42\x40\x40\x42\x52\x00\x00\x53\x73\xc0\x1f\x0d\x40\x40\x40\x40\x40\x40\x40\x40\x83\x00\x00\x00\x00\x00\x00\x00\x00\x83\x00\x00\x00\x00\x00\x00\x00\x00\x40\x52\x00\x40'
-        bytes = ffi.new('char []', data)
-        size = ffi.new('size_t *', ffi.sizeof(bytes))
+
         msg3 = Message()
-        msg3.decode(bytes, size)
+        msg3.decode(data)
         assert msg3.priority == 4, (msg3.priority)
 
         # Minified message with zero length HEADER and PROPERTIES lists
         data = b'\x00\x53\x70\x45' b'\x00\x53\x73\x45'
-        bytes = ffi.new('char []', data)
-        size = ffi.new('size_t *', ffi.sizeof(bytes))
         msg4 = Message()
-        msg4.decode(bytes, size)
+        msg4.decode(data)
         assert msg4.priority == 4, (msg4.priority)

